edgecontroller panics when there is no configuration for cloudhub.ca, cloudhub.cert, cloudhub.key.

**What happened**:
edgecontroller crashes with missing configuration for cloudhub.ca, cloudhub.cert, cloudhub.key.

**What you expected to happen**:
Exit gracefully with missing configuration information.

**How to reproduce it (as minimally and precisely as possible)**:
Remove configuration for
  ca: /tmp/rootCA.crt
  cert: /tmp/edge.crt
  key: /tmp/edge.key
in controller.yaml.

**Anything else we need to know?**:
Also, need to remove the code in cloudhub.go:Start() for redundant code.

	// init filter
	pool := x509.NewCertPool()
	ok := pool.AppendCertsFromPEM(util.HubConfig.Ca)
	if !ok {
		panic(fmt.Errorf("fail to load ca content"))
	}

**Environment**:
- KubeEdge version: 0.2
- Hardware configuration: Virtual box vm - 2 vcpus / 2GB memory
- OS (e.g. from /etc/os-release): Ubuntu 18.04.1 LTS
- Kernel (e.g. `uname -a`):Linux 4.18.0-16-generic 
- Others:

Logs==>

2019/04/04 13:30:25 WARN: empty configurtion from [CommandlineSource]
2019/04/04 13:30:25 WARN: empty configurtion from [MemorySource]
New file source added for configuration: /opt/spire/app-binaries/conf/controller.yamlNew file source added for configuration: /opt/spire/app-binaries/conf/edge.yamlNew file source added for configuration: /opt/spire/app-binaries/conf/logging.yamlNew file source added for configuration: /opt/spire/app-binaries/conf/modules.yaml/opt/spire/app-binaries/conf/logging.yamllogConfig:&{INFO edgecontroller.log [file stdout] true false  }
2019/04/04 13:30:25 RollingPolicy is empty, use default policy[size]
2019-04-04 13:30:25.959 +05:30 INFO core/module.go:41 module cloudhub registered
2019-04-04 13:30:25.981 +05:30 INFO config/buffer.go:27 update pod status buffer: 1024
2019-04-04 13:30:25.981 +05:30 INFO config/buffer.go:34 Update node status buffer: 1024
2019-04-04 13:30:25.981 +05:30 INFO config/buffer.go:41 query config map buffer: 1024
2019-04-04 13:30:25.981 +05:30 INFO config/buffer.go:48 query secret buffer: 1024
2019-04-04 13:30:25.981 +05:30 INFO config/context.go:28  send module name: cloudhub
2019-04-04 13:30:25.981 +05:30 INFO config/context.go:35 receive module name: controller
2019-04-04 13:30:25.981 +05:30 INFO config/context.go:42 response module name: cloudhub
2019-04-04 13:30:25.981 +05:30 INFO config/kube.go:38 kube master: http://localhost:8080
2019-04-04 13:30:25.981 +05:30 INFO config/kube.go:45 kube config: 
2019-04-04 13:30:25.981 +05:30 INFO config/kube.go:52 kube namespace: 
2019-04-04 13:30:25.981 +05:30 INFO config/kube.go:59 kube content type: application/vnd.kubernetes.protobuf
2019-04-04 13:30:25.981 +05:30 INFO config/kube.go:66 kube QPS: 5.000000
2019-04-04 13:30:25.981 +05:30 INFO config/kube.go:73 kube burst: 10
2019-04-04 13:30:25.982 +05:30 INFO config/kube.go:80 kube update frequency: 10s
2019-04-04 13:30:25.982 +05:30 INFO config/load.go:27 update pod status workers: 1
2019-04-04 13:30:25.982 +05:30 INFO config/load.go:34 update node status workers: 1
2019-04-04 13:30:25.982 +05:30 INFO config/load.go:41 query config map workers: 4
2019-04-04 13:30:25.982 +05:30 INFO config/load.go:48 query secret workers: 4
2019-04-04 13:30:25.982 +05:30 INFO config/messagelayer.go:18 message layer: context
2019-04-04 13:30:26.020 +05:30 INFO core/module.go:41 module controller registered
2019-04-04 13:30:26.021 +05:30 INFO core/core.go:23 starting module cloudhub
panic: interface conversion: interface {} is nil, not string

goroutine 9 [running]:
github.com/kubeedge/kubeedge/cloud/edgecontroller/pkg/cloudhub.(*cloudHub).Start(0xc0000b4130, 0xc00030aca0)
	/home/root1/gopath/src/github.com/kubeedge/kubeedge/cloud/edgecontroller/pkg/cloudhub/cloudhub.go:39 +0x4fd
created by github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/beehive/pkg/core.StartModules
	/home/root1/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/beehive/pkg/core/core.go:22 +0x153

--- a/cloud/pkg/cloudhub/cloudhub.go
+++ b/cloud/pkg/cloudhub/cloudhub.go
@@ -2,8 +2,10 @@ package cloudhub
 
 import (
 	"io/ioutil"
+	"os"
 
 	"github.com/kubeedge/beehive/pkg/common/config"
+	"github.com/kubeedge/beehive/pkg/common/log"
 	"github.com/kubeedge/beehive/pkg/core"
 	"github.com/kubeedge/beehive/pkg/core/context"
 	"github.com/kubeedge/kubeedge/cloud/pkg/cloudhub/channelq"
@@ -33,16 +35,19 @@ func (a *cloudHub) Start(c *context.Context) {
 
 	cafile, err := config.CONFIG.GetValue("cloudhub.ca").ToString()
 	if err != nil {
+		log.LOGGER.Info("missing cloudhub.ca configuration key, loading default path and filename ./" + chconfig.DefaultCAFile)
 		cafile = chconfig.DefaultCAFile
 	}
 
 	certfile, err := config.CONFIG.GetValue("cloudhub.cert").ToString()
 	if err != nil {
+		log.LOGGER.Info("missing cloudhub.cert configuration key, loading default path and filename ./" + chconfig.DefaultCertFile)
 		certfile = chconfig.DefaultCertFile
 	}
 
 	keyfile, err := config.CONFIG.GetValue("cloudhub.key").ToString()
 	if err != nil {
+		log.LOGGER.Info("missing cloudhub.key configuration key, loading default path and filename ./" + chconfig.DefaultKeyFile)
 		keyfile = chconfig.DefaultKeyFile
 	}
 
@@ -61,14 +66,19 @@ func (a *cloudHub) Start(c *context.Context) {
 		errs = append(errs, err.Error())
 	}
 
-	if len(errs) <= 0 {
+	if len(errs) > 0 {
 		//  TBD : add code for to sync graceful exit of modules
-		return
+		log.LOGGER.Errorf("cloudhub failed with errors : %v", errs)
+		os.Exit(1)
 	}
 
 	eventq, err := channelq.NewChannelEventQueue(c)
 	// start the cloudhub server
-	wsserver.StartCloudHub(util.HubConfig, eventq)
+	err = wsserver.StartCloudHub(util.HubConfig, eventq)
+	if err != nil {
+		log.LOGGER.Errorf("cloudhub start failed with errors : %v", err)
+		os.Exit(1)
+	}
 	wsserver.EventHandler.Context = c
 	stopchan := make(chan bool)
 	<-stopchan

--- a/cloud/pkg/cloudhub/cloudhub.go
+++ b/cloud/pkg/cloudhub/cloudhub.go
@@ -67,7 +67,6 @@ func (a *cloudHub) Start(c *context.Context) {
 	}
 
 	if len(errs) > 0 {
-		//  TBD : add code for to sync graceful exit of modules
 		log.LOGGER.Errorf("cloudhub failed with errors : %v", errs)
 		os.Exit(1)
 	}

--- a/cloud/pkg/cloudhub/cloudhub.go
+++ b/cloud/pkg/cloudhub/cloudhub.go
@@ -1,8 +1,6 @@
 package cloudhub
 
 import (
-	"crypto/x509"
-	"fmt"
 	"io/ioutil"
 
 	"github.com/kubeedge/beehive/pkg/common/config"
@@ -10,6 +8,7 @@ import (
 	"github.com/kubeedge/beehive/pkg/core/context"
 	"github.com/kubeedge/kubeedge/cloud/pkg/cloudhub/channelq"
 	"github.com/kubeedge/kubeedge/cloud/pkg/cloudhub/common/util"
+	chconfig "github.com/kubeedge/kubeedge/cloud/pkg/cloudhub/config"
 	"github.com/kubeedge/kubeedge/cloud/pkg/cloudhub/wsserver"
 )
 
@@ -31,30 +30,40 @@ func (a *cloudHub) Group() string {
 
 func (a *cloudHub) Start(c *context.Context) {
 	a.context = c
-	var err error
-	caI := config.CONFIG.GetConfigurationByKey("cloudhub.ca")
-	certI := config.CONFIG.GetConfigurationByKey("cloudhub.cert")
-	keyI := config.CONFIG.GetConfigurationByKey("cloudhub.key")
 
-	util.HubConfig.Ca, err = ioutil.ReadFile(caI.(string))
+	cafile, err := config.CONFIG.GetValue("cloudhub.ca").ToString()
 	if err != nil {
-		panic(err)
+		cafile = chconfig.DefaultCAFile
 	}
 
-	util.HubConfig.Cert, err = ioutil.ReadFile(certI.(string))
+	certfile, err := config.CONFIG.GetValue("cloudhub.cert").ToString()
 	if err != nil {
-		panic(err)
-	}
-	util.HubConfig.Key, err = ioutil.ReadFile(keyI.(string))
-	if err != nil {
-		panic(err)
+		certfile = chconfig.DefaultCertFile
 	}
 
-	// init filter
-	pool := x509.NewCertPool()
-	ok := pool.AppendCertsFromPEM(util.HubConfig.Ca)
-	if !ok {
-		panic(fmt.Errorf("fail to load ca content"))
+	keyfile, err := config.CONFIG.GetValue("cloudhub.key").ToString()
+	if err != nil {
+		keyfile = chconfig.DefaultKeyFile
+	}
+
+	errs := make([]string, 0)
+
+	util.HubConfig.Ca, err = ioutil.ReadFile(cafile)
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+	util.HubConfig.Cert, err = ioutil.ReadFile(certfile)
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+	util.HubConfig.Key, err = ioutil.ReadFile(keyfile)
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+
+	if len(errs) <= 0 {
+		//  TBD : add code for to sync graceful exit of modules
+		return
 	}
 
 	eventq, err := channelq.NewChannelEventQueue(c)

--- a/cloud/pkg/cloudhub/config/default.go
+++ b/cloud/pkg/cloudhub/config/default.go
@@ -1,0 +1,8 @@
+package config
+
+// Default constants definitions for cloudhub configuration
+const (
+	DefaultCAFile   = "cloudhubca.crt"
+	DefaultCertFile = "cloudhub.crt"
+	DefaultKeyFile  = "cloudhub.key"
+)

--- a/cloud/pkg/cloudhub/wsserver/server.go
+++ b/cloud/pkg/cloudhub/wsserver/server.go
@@ -370,16 +370,16 @@ func (f *FilterWriter) Write(p []byte) (n int, err error) {
 }
 
 // StartCloudHub starts the cloud hub service
-func StartCloudHub(config *util.Config, eventq *channelq.ChannelEventQueue) {
+func StartCloudHub(config *util.Config, eventq *channelq.ChannelEventQueue) error {
 	// init certificate
 	pool := x509.NewCertPool()
 	ok := pool.AppendCertsFromPEM(config.Ca)
 	if !ok {
-		panic(fmt.Errorf("fail to load ca content"))
+		return fmt.Errorf("fail to load ca content")
 	}
 	cert, err := tls.X509KeyPair(config.Cert, config.Key)
 	if err != nil {
-		panic(err)
+		return err
 	}
 	tlsConfig := tls.Config{
 		ClientCAs:    pool,
@@ -412,4 +412,6 @@ func StartCloudHub(config *util.Config, eventq *channelq.ChannelEventQueue) {
 	}
 	bhLog.LOGGER.Infof("Start cloud hub service")
 	go s.ListenAndServeTLS("", "")
+
+	return nil
 }

--- a/cloud/pkg/controller/controller.go
+++ b/cloud/pkg/controller/controller.go
@@ -37,7 +37,7 @@ func (ctl *Controller) Start(c *bcontext.Context) {
 	config.Context = c
 	upstream, err := controller.NewUpstreamController()
 	if err != nil {
-		log.LOGGER.Warnf("new upstream controller failed with error: %s", err)
+		log.LOGGER.Errorf("new upstream controller failed with error: %s", err)
 		os.Exit(1)
 	}
 	upstream.Start()

--- a/cloud/pkg/devicecontroller/module.go
+++ b/cloud/pkg/devicecontroller/module.go
@@ -1,6 +1,7 @@
 package devicecontroller
 
 import (
+	"os"
 	"time"
 
 	"github.com/kubeedge/beehive/pkg/common/log"
@@ -39,14 +40,15 @@ func (dctl *DeviceController) Start(c *bcontext.Context) {
 	dctl.stopChan = make(chan bool)
 	downstream, err := controller.NewDownstreamController()
 	if err != nil {
-		log.LOGGER.Warnf("New downstream controller failed with error: %s", err)
-		return
+		log.LOGGER.Errorf("New downstream controller failed with error: %s", err)
+		os.Exit(1)
 	}
 	upstream, err := controller.NewUpstreamController(downstream)
 	if err != nil {
-		log.LOGGER.Warnf("new upstream controller failed with error: %s", err)
-		return
+		log.LOGGER.Errorf("new upstream controller failed with error: %s", err)
+		os.Exit(1)
 	}
+
 	downstream.Start()
 	// wait for downstream controller to start and load deviceModels and devices
 	time.Sleep(1 * time.Second)

--- a/edge/pkg/edged/edged.go
+++ b/edge/pkg/edged/edged.go
@@ -3,11 +3,13 @@ package edged
 import (
 	"encoding/json"
 	"fmt"
-	"github.com/kubeedge/kubeedge/edge/pkg/common/modules"
+	"os"
 	"sync"
 	"time"
 
-	"k8s.io/api/core/v1"
+	"github.com/kubeedge/kubeedge/edge/pkg/common/modules"
+
+	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/clock"
 	"k8s.io/apimachinery/pkg/util/sets"
@@ -169,12 +171,12 @@ func (e *edged) Start(c *context.Context) {
 	e.context = c
 	if err := e.initializeModules(); err != nil {
 		log.LOGGER.Errorf("initialize module error: %v", err)
-		return
+		os.Exit(1)
 	}
 	err := e.makePodDir()
 	if err != nil {
 		log.LOGGER.Errorf("create pod dir [%s] failed: %v", e.getPodsDir(), err)
-		return
+		os.Exit(1)
 	}
 	e.metaClient = metaclient.New(c)
 	e.statusManager = status.NewManager(e.kubeClient, e.podManager, utilpod.NewPodDeleteSafety(), e.metaClient)

--- a/edge/pkg/eventbus/event_bus.go
+++ b/edge/pkg/eventbus/event_bus.go
@@ -3,6 +3,7 @@ package eventbus
 import (
 	"encoding/json"
 	"fmt"
+	"os"
 
 	"github.com/256dpi/gomqtt/packet"
 	"github.com/kubeedge/beehive/pkg/common/config"
@@ -56,7 +57,8 @@ func (eb *eventbus) Start(c *context.Context) {
 
 	nodeID := config.CONFIG.GetConfigurationByKey("edgehub.controller.node-id")
 	if nodeID == nil {
-		panic("node id not configured")
+		log.LOGGER.Errorf("node id not configured")
+		os.Exit(1)
 	}
 
 	mqttBus.NodeID = nodeID.(string)
@@ -99,14 +101,16 @@ func (eb *eventbus) Start(c *context.Context) {
 		}
 
 		if qos.(int) < int(packet.QOSAtMostOnce) || qos.(int) > int(packet.QOSExactlyOnce) || sessionQueueSize.(int) <= 0 {
-			panic("mqtt.qos must be one of [0,1,2] or mqtt.session-queue-size must > 0")
+			log.LOGGER.Errorf("mqtt.qos must be one of [0,1,2] or mqtt.session-queue-size must > 0")
+			os.Exit(1)
 		}
 		// launch an internal mqtt server only
 		mqttServer = mqttBus.NewMqttServer(sessionQueueSize.(int), internalMqttURL.(string), retain.(bool), qos.(int))
 		mqttServer.InitInternalTopics()
 		err := mqttServer.Run()
 		if err != nil {
-			panic(fmt.Sprintf("Launch mqtt broker failed, %s", err.Error()))
+			log.LOGGER.Errorf("Launch mqtt broker failed, %s", err.Error())
+			os.Exit(1)
 		}
 	}
 

